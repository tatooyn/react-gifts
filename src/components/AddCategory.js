import React,{useState} from 'react';
import PropTypes from 'prop-types'
// ComPadre: "setcategorias()": 
// const [categorias, setcategorias] = useState(['One Piece']);
const AddCategory= ({setcategorias=()=>{}})=>{
    const [nuevaCategoria, setnuevaCategoria] = useState('');

    const enviar=(e)=>{
        e.preventDefault();
        if(nuevaCategoria.trim().length>2){
            setcategorias(cats=>[nuevaCategoria,...cats]);
            setnuevaCategoria('');
        }
    }
    return <>
        <form onSubmit={enviar}>
            <div className='form-group'>
                <label>Añade una Categoria</label>
                <input value={nuevaCategoria}
                    onChange={e=>setnuevaCategoria(e.target.value)}
                    className='form-control'
                    placeholder='Añade cualquier palabra o titulo'/>
                    {nuevaCategoria}
                <button type='submit'>Añadir</button>
            </div>
        </form>
    </>;
};
AddCategory.propTypes={
    setcategorias: PropTypes.func.isRequired
}
export default AddCategory;