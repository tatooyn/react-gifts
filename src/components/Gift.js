import React from 'react';
import PropTypes from 'prop-types';
const Gift= ({id='', title, url})=>{
    return <div className='card animate__animated animate__fadeInDown'>
        <h3>{title}</h3>
        <img src={url} alt={id}/>
    </div>;
};
Gift.propTypes={
    title: PropTypes.string.isRequired,
    url: PropTypes.string.isRequired,
}
export default Gift;