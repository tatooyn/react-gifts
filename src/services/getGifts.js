const getGifts=async (category)=>{
    const api_key='7wCkTnLjyUVjC8Ua0Vt3usELbsfXL81w';
    const q=encodeURI(category);
    const url=`https://api.giphy.com/v1/gifs/search?api_key=${api_key}&q=${q}&limit=5`;
    const {data}=await fetch(url).then(res=>res.json());
    const gifts=data.map(img=>({
        id:img.id,
        title:img.title,
        url:img.images?.downsized_medium.url
    }))
    return gifts;
};
export default getGifts;