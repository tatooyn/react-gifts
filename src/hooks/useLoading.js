import { useState,useEffect } from "react";
import getGifts from '../services/getGifts';

const useLoading=(category)=>{
    const [data, setdata] = useState({
        data:[],
        loading:true,
    });
    useEffect(() => {
        getGifts(category)
            .then(img=>setdata({data:img,loading:false}));
    }, [category]);
    return data;
}
export default useLoading;