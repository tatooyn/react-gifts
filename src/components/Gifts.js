import React from 'react';
import PropTypes from 'prop-types';
import Gift from './Gift';
import useLoading from '../hooks/useLoading';


const Gifts= ({category=''})=>{
    const {data:image, loading}=useLoading(category);

    const imagesRender=image.map(img=><Gift key={img.id} {...img}/>);
    return <>
        {loading && <h3>loading</h3>}
        <h1>{category}</h1>
        <div className='card-grid'>
            {imagesRender}
        </div>
    </>;
};
Gifts.popTypes={
    category: PropTypes.string.isRequired
}
export default Gifts;