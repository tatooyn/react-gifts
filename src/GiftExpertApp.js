import React, { useState } from 'react';
import AddCategory from './components/AddCategory';
import Gifts from './components/Gifts';
const GiftExpertApp= ({initCategories=[]})=>{
    const [categorias, setcategorias] = useState(initCategories);
    const categoriasRender=categorias.map(categoria=>{
        return <Gifts key={categoria} category={categoria}/>;
    });
    // const addCategoria=(nuevaCategoria)=>{
    //     setcategorias(cat=>[...categorias,nuevaCategoria])
    // };
    return <>
        <h2>GiftExpertApp</h2>
        <hr/>
        <AddCategory setcategorias={setcategorias}/>
        <ol>
            {categoriasRender}
        </ol>
        <br/><br/>
        <p>Aplicación desarrollada con React, por Andrés Cardona Alvarez</p>
        <p>Del curso "React: De cero a experto ( Hooks y MERN ); de Udemy"</p>
        <p>GitHub: <a href='https://github.com/tatooyn/react-gifts.git'>https://github.com/tatooyn/react-gifts.git</a> </p>
        <h4>Temas:</h4>
        <ul>
            <li>Creacion y comunicacion entre componentes con Props</li>
            <li>UseState, UseEfect, CustomHooks</li>
            <li>Test Unitatios, Enzyme, react-hooks-testing</li>
        </ul>
    </>;
};
export default GiftExpertApp;