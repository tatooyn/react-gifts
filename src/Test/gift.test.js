import React from 'react';
import { shallow } from 'enzyme';
import {renderHook} from '@testing-library/react-hooks';
import Gift from '../components/Gift';
import AddCategory from '../components/AddCategory';
import getGifts from '../services/getGifts';
import useLoading from '../hooks/useLoading';
import Gifts from '../components/Gifts';
import GiftExpertApp from '../GiftExpertApp';

jest.mock('../hooks/useLoading');

describe('== Gift component ==',()=>{
    // <Gift/>
    const props={
        id:1,
        title:'one piece',
        url:'htttp://blabla'}
    const giftComp=shallow(<Gift {...props}/>);
    
    
    test('°Gift snapshot',()=>{
        expect(giftComp).toMatchSnapshot();
    })
    test('°Gift props',()=>{
        const propsTest={
            id:giftComp.find('img').prop('alt'),
            title:giftComp.find('h3').text(),
            url:giftComp.find('img').prop('src')
        }
        expect(propsTest).toEqual(props);
    })
    test('°Fetch Gift=',async ()=>{
        const data=await getGifts('one piece');
        expect(data.length).toBe(5);
    })
})
describe('== AddGift component ==',()=>{
    // <AddCategory/>
    const setCategorias=()=>{};
    let giftAddComp=shallow(<AddCategory setcategorias={setCategorias}/>);
    beforeEach(()=>{
        giftAddComp=shallow(<AddCategory setcategorias={setCategorias}/>);
    })
    
    test('°snapshot',()=>{
        expect(giftAddComp).toMatchSnapshot();
    })
    test('°input',()=>{
        const input=giftAddComp.find('input');
        input.simulate('change',{target:{value:'Batman'}});
    })
    test('°form submit',()=>{
        const form=giftAddComp.find('form');
        form.simulate('submit',{preventDefault(){}});
        expect(setCategorias).not.toHaveBeenCalled;
    })
    test('°setCategory and cleanInput',()=>{
        const input=giftAddComp.find('input');
        input.simulate('change',{target:{value:'Batman'}});
        giftAddComp.find('form').simulate('submit',{preventDefault(){}});
        const inputValue=input.prop('value');
        expect(setCategorias).toHaveBeenCalled;
        expect(inputValue).toBe('');
    })
})

describe('== Gifts component ==',()=>{

    useLoading.mockReturnValue({
        data:[],
        loading:true,
    });
    const giftsComp=shallow(<Gifts category='mononoke'/>);
    test('°Gifts snapshot empty',()=>{
        expect(giftsComp).toMatchSnapshot();
    })
    test('°Gifts snapshot with gift',()=>{
        const gifts=[{id:'a',url:'http',title:'a'},
            {id:'a',url:'http',title:'a'}]
        useLoading.mockReturnValue({
            data:gifts,
            loading:false,
        });
        const giftsComp=shallow(<Gifts category='mononoke'/>);
        expect(giftsComp.find('h3').exists()).toBe(false);
    })
    test('°Gifts img empty',()=>{
        const imgs=giftsComp.filter('img').length;
        expect(imgs).toBe(0);
    })
    test('°Gifts img images',()=>{
        const gifts=[{id:'a',url:'http',title:'a'},
            {id:'a',url:'http',title:'a'}]
        useLoading.mockReturnValue({
            data:gifts,
            loading:false,
        });
        const giftsComp=shallow(<Gifts category='mononoke'/>);
        const imgs=giftsComp.find('Gift').length;
        expect(imgs).toBe(2);
    })
})
describe('== gifts App ==',()=>{
    test('°snaopshot',()=>{
        const giftAppComp=shallow(<GiftExpertApp/>);
        expect(giftAppComp).toMatchSnapshot();
    })
    test('° Number of Gifts',()=>{
        const giftAppComp=shallow(<GiftExpertApp/>);
        expect(giftAppComp.find('Gifts').length).toBe(0);
    })
    test('° Number of Gifts',()=>{
        const initCategories=['one piece', 'batman'];
        const giftAppComp=shallow(<GiftExpertApp initCategories={initCategories}/>);
        expect(giftAppComp).toMatchSnapshot();
        expect(giftAppComp.find('Gifts').length).toBe(2);
    })
})

