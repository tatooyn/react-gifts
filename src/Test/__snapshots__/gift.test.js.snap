// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`== AddGift component == °snapshot 1`] = `
<Fragment>
  <form
    onSubmit={[Function]}
  >
    <div
      className="form-group"
    >
      <label>
        Nueva Categoria
      </label>
      <input
        className="form-control"
        onChange={[Function]}
        value=""
      />
    </div>
  </form>
</Fragment>
`;

exports[`== Gift component == °Gift snapshot 1`] = `
<div
  className="card animate__animated animate__fadeInDown"
>
  <h3>
    one piece
  </h3>
  <img
    alt={1}
    src="htttp://blabla"
  />
</div>
`;

exports[`== Gifts component == °Gifts snapshot empty 1`] = `
<Fragment>
  <h3>
    loading
  </h3>
  <h1>
    mononoke
  </h1>
  <div
    className="card-grid"
  />
</Fragment>
`;

exports[`== gifts App == ° Number of Gifts 1`] = `
<Fragment>
  <h2>
    GiftExpertApp
  </h2>
  <hr />
  <AddCategory
    setcategorias={[Function]}
  />
  <ol>
    <Gifts
      category="one piece"
      key="one piece"
    />
    <Gifts
      category="batman"
      key="batman"
    />
  </ol>
</Fragment>
`;

exports[`== gifts App == °snaopshot 1`] = `
<Fragment>
  <h2>
    GiftExpertApp
  </h2>
  <hr />
  <AddCategory
    setcategorias={[Function]}
  />
  <ol />
</Fragment>
`;
